### PR TITLE
Try reading data from underlying source if the data has been corrupted

### PR DESCRIPTION
**Pull request by Praveen2112**
## Description
Fixes https://github.com/trinodb/trino/issues/21342 and https://github.com/trinodb/trino/issues/21147.

These exists a race condition in Alluxio cache which occurs when two different threads are trying to read a same file, 
then it fails to read the data file with the following message

```
	Caused by: java.io.IOException: Read -1 bytes from cache
		at io.trino.filesystem.alluxio.AlluxioInputHelper.doInternalCacheRead(AlluxioInputHelper.java:120)
		at io.trino.filesystem.alluxio.AlluxioInputHelper.lambda$doCacheRead$0(AlluxioInputHelper.java:89)
		at io.trino.filesystem.alluxio.AlluxioTracing.withTracing(AlluxioTracing.java:29)
		at io.trino.filesystem.alluxio.AlluxioInputHelper.doCacheRead(AlluxioInputHelper.java:87)
		at io.trino.filesystem.alluxio.AlluxioInput.readFully(AlluxioInput.java:79)
		at io.trino.filesystem.alluxio.AlluxioInput.readTail(AlluxioInput.java:128)
		at io.trino.filesystem.TrinoInput.readTail(TrinoInput.java:43)
		at io.trino.filesystem.tracing.TracingInput.lambda$readTail$3(TracingInput.java:81)
		at io.trino.filesystem.tracing.Tracing.withTracing(Tracing.java:47)
		at io.trino.filesystem.tracing.TracingInput.readTail(TracingInput.java:81)
		at io.trino.plugin.hive.parquet.MemoryParquetDataSource.<init>(MemoryParquetDataSource.java:56)
		at io.trino.plugin.hive.parquet.ParquetPageSourceFactory.createDataSource(ParquetPageSourceFactory.java:319)
		at io.trino.plugin.hive.parquet.ParquetPageSourceFactory.createPageSource(ParquetPageSourceFactory.java:222)
		... 20 more
```

CacheMaanger#put doesn't guarantee that file is written by Alluxio and it is ready to be read. From the CacheManager Javadocs

```
Puts a page into the cache manager. This method is best effort. It is possible that this put operation returns without page written.
```

## Additional context and related issues

We could reproduce this issue by increasing the repeated test count in `TestIcebergMinioParquetCachingConnectorSmokeTest#testDeleteRowsConcurrently`

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Review by findepi — APPROVED** [review]
is this testable?

**Review by jkylling — APPROVED** [review]

**Comment by Praveen2112** [comment]
> is this testable?

This is seen during the concurrent reads. We can try to run the tests locally by increasing the test invocation count for `TestIcebergMinioParquetCachingConnectorSmokeTest#testDeleteRowsConcurrently` without this change it would fail for random test runs.

File: lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java [hunk 1]
```diff
@@ -113,12 +110,9 @@ private int doInternalCacheRead(long position, byte[] bytes, int offset, int len
         int remainingLength = length;
         while (remainingLength > 0) {
             int bytesReadFromCache = readPageFromCache(position, bytes, offset, remainingLength);
-            if (bytesReadFromCache == 0) {
+            if (bytesReadFromCache <= 0) {
                 break;
```
Comment (findinpath) [inline lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java:114]: pre-existing: 

https://github.com/trinodb/trino/blob/0c1fdcb2a736673fad219ea34a7f16dc70104582/lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java#L127

Are we supposed to record the cache when not reading anything from it?

**Review by findinpath — APPROVED** [review]

File: lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java (hunk 1 above)
Comment (Praveen2112) [reply lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java:114]: Some parts of the cache could be read right ? `bytesRead` would capture that.

**Comment by findepi** [comment]
> > is this testable?
> 
> This is seen during the concurrent reads. We can try to run the tests locally by increasing the test invocation count for `TestIcebergMinioParquetCachingConnectorSmokeTest#testDeleteRowsConcurrently` without this change it would fail for random test runs.

I know this, but that's not the answer i am looking for.

File: lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java [hunk 2]
```diff
@@ -113,12 +110,9 @@ private int doInternalCacheRead(long position, byte[] bytes, int offset, int len
         int remainingLength = length;
         while (remainingLength > 0) {
             int bytesReadFromCache = readPageFromCache(position, bytes, offset, remainingLength);
-            if (bytesReadFromCache == 0) {
+            if (bytesReadFromCache <= 0) {
```
Comment (raunaqmorarka) [inline lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java:113]: Can we include a code comment explaining why ?

**Review by raunaqmorarka — APPROVED** [review]

File: lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java (hunk 2 above)
Comment (Praveen2112) [reply lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java:115]: We have added it.

**Comment by Praveen2112** [comment]
@findepi We have added a test case where we try to mimic the concurrent access behaviour manually by dropping the pages and rewriting them externally.

File: lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java [hunk 3]
```diff
@@ -113,12 +110,11 @@ private int doInternalCacheRead(long position, byte[] bytes, int offset, int len
         int remainingLength = length;
         while (remainingLength > 0) {
             int bytesReadFromCache = readPageFromCache(position, bytes, offset, remainingLength);
-            if (bytesReadFromCache == 0) {
+            // When dealing with concurrent access for a new file, CacheManager#put doesn't guarantee the page is fully written,
+            // and trying to access the page could return -1, so we retry reading data
```
Comment (raunaqmorarka) [inline lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java:114]: by "retry" do you mean the retry from cache or fallback to remote source ? i'm assuming it's the latter, would be good to clarify that

Comment (Praveen2112) [reply lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java]: Updated them

File: lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java [hunk 4]
```diff
@@ -24,7 +24,6 @@
 import io.opentelemetry.api.trace.Tracer;
 import io.trino.filesystem.Location;
 
-import java.io.IOException;
```
Comment (findepi) [inline lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioInputHelper.java:27]: > Try reading data from underlying source if the data has been corrupted

Try reading -> Read

**Review by findepi — APPROVED** [review]

**Pull request merged**